### PR TITLE
electrum: add support for LedgerHQ hardware wallet

### PR DESCRIPTION
**Pull request by k0001**
###### Motivation for this change

Adding support for LedgerHQ hardware wallet in electrum by adding a dependency on `btchip`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by k0001** [comment]
Could this be backported to 17.09 as well? Thanks.

File: pkgs/development/python-modules/btchip/default.nix [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    inherit pname version;
+    sha256 = "16g9l3rpxpvvkdx08mgy0ligvsfcpzdrh4hplj104cpprrbsqd6v";
+  };
+
+  buildInputs = [ hidapi pyscard ecdsa ];
+
+  meta = with stdenv.lib; {
+    description = "Python communication library for Ledger Hardware Wallet products";
+    homepage = "https://github.com/LedgerHQ/btchip-python";
+    license = licenses.apache2;
```
Comment (joachifm) [inline pkgs/development/python-modules/btchip/default.nix]: Use `licenses.asl20`. You can test meta eval by `nix-instantiate --eval --strict -A foo.meta`.

**Review by joachifm — CHANGES_REQUESTED** [review]

**Comment by geistesk** [comment]
Is it necessary to add this plugin for each electrum-instance even if the majority won't use it? I feel quite uncomfortable to add unnecessary plugins in a Bitcoin-client.
Perhaps you can disable it by default and let the user enable it by overriding a list of additional plugins?

**Comment by k0001** [comment]
@joachifm fixed.

@geistesk I don't know. An argument can be made in favor of having a fully functional Electrum out of the box. `btchip` doesn't have that many dependencies anyway, so the number of dependencies is presumably not a problem. Without `btchip`, the current situation is that if you try to link a hardware wallet from Electrum it just fails without any useful error message.

**Comment by k0001** [comment]
@geistesk also, the Electrum we ship in nixpkgs already supports the Trezor hardware wallet.

File: pkgs/development/python-modules/btchip/default.nix [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  pname = "btchip-python";
+  version = "0.1.21";
+  name = "${pname}-${version}";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "16g9l3rpxpvvkdx08mgy0ligvsfcpzdrh4hplj104cpprrbsqd6v";
+  };
+
+  buildInputs = [ hidapi pyscard ecdsa ];
```
Comment (FRidh) [inline pkgs/development/python-modules/btchip/default.nix]: these are not needed during runtime?

Comment (k0001) [reply pkgs/development/python-modules/btchip/default.nix]: @FRidh there are definitely references to these three libraries in the `btchip` code.

Comment (FRidh) [reply pkgs/development/python-modules/btchip/default.nix]: in order for Python modules to find each other, they need to be added as `propagatedBuildInputs`.

**Comment by joachifm** [comment]
I'd be fine with making plugins a parameter to the package expression but we can do that in a separate feature.

**Comment by geistesk** [comment]
@k0001 I fully understand the benefits of a "batteries-included" client. However, on the other side there will be a lot of additional code with possible security-flows in a Bitcoin-client (or any software at all). That's why I'd love to see an overrideable list of possible plugins.

@joachifm If we are going to introduce this feature later, we will probably end up with some "legacy"-dependencies. But I'd like to look into this subject and eventually create a PR.

**Comment by nyanloutre** [comment]
Is this still on hold ? What work needs to be done for this to be merged ?

**Comment by joachifm** [comment]
Seems to me that @FRidh's feedback has not been adressed

**Pull request closed without merging**